polynote-kernel/src/main/scala/polynote/messages/Messages.scala

```diff
@@ -127,47 +125,175 @@ object RunCell extends MessageCompanion[RunCell](3)
-      copy(pos = pos + otherContent.length - otherDeleteLength)
+sealed trait ContentEdit {
+  def pos: Int
+  def lengthDelta: Int
 
-    // if the other edit is entirely after this edit, nothing to do
-    case ContentEdit(otherPos, _, _) if otherPos >= pos + deleteLength => this
+  def rebase(other: ContentEdits): ContentEdits = ContentEdits(ShortList(ContentEdit.rebaseAll(this, other.edits)._1))
 
-    // if the other edit affects the region that was replaced by this edit, then they conflict
-    // we want the minimal edit that replaces the entire conflict region
-    case ContentEdit(otherPos, otherDeleteLength, otherContent) =>
+  def applyTo(rope: Rope): Rope = this match {
+    case Insert(pos, content) => rope.insertAt(pos, Rope(content))
+    case Delete(pos, length)  => rope.delete(pos, length)
+  }
 
-      // start at the earlier position
-      val start = math.min(pos, otherPos)
+  def nonEmpty: Boolean
+}
 
-      // They deleted to this position
-      val theirTarget = otherPos + otherDeleteLength
+final case class Insert(pos: Int, content: String) extends ContentEdit {
+  def lengthDelta: Int = content.length
 
-      // I would have deleted up to this position...
-      val myTarget = pos + deleteLength
+  def nonEmpty: Boolean = content.nonEmpty
 
-      // So I should delete this much extra
-      val deleteExtra = math.max(0, myTarget - theirTarget) + math.max(0, otherPos - pos)
+  override def toString: String = s"""Insert($pos, "$content")"""
+}
 
-      val combinedContent = if (otherPos <= pos) otherContent + content else content + otherContent
+object Insert {
+  implicit val insertTag: Discriminator[ContentEdit, Insert, Byte] = Discriminator(0)
+}
+
+final case class Delete(pos: Int, length: Int) extends ContentEdit {
+  override def lengthDelta: Int = -length
+
+  override def nonEmpty: Boolean = length != 0
+}
 
-      ContentEdit(start, otherContent.length + deleteExtra, combinedContent)
+object Delete {
+  implicit val deleteTag: Discriminator[ContentEdit, Delete, Byte] = Discriminator(1)
+}
+
+object ContentEdit {
+  implicit val discriminated: Discriminated[ContentEdit, Byte] = Discriminated(byte)
+
+  // rebase a onto b and b onto a
+  def rebase(a: ContentEdit, b: ContentEdit): (List[ContentEdit], List[ContentEdit]) = (a, b) match {
+
+    // if A is before A, or A and B are at the same spot but (A is shorter than B or equal in length but lexically before B)
+    // then A comes before B.
+    case (winner @ Insert(pos1, content1), Insert(pos2, content2)) if
+      (pos1 < pos2) ||
+        (pos1 == pos2 &&
+          (content1.length < content2.length || (content1.length == content2.length && content1 < content2))) =>
+            List(winner) -> List(Insert(pos2 + content1.length, content2))
+
+    // Otherwise insert B comes before insert A
+    case (Insert(pos1, content1), winner @ Insert(_, content2)) =>
```

Comment:
It might be a little clearer to use A and B in the code as well as the comments - e.g., `case (Insert(posA, contentA), winner @ Insert(_, contentB))`

(Or use 1 and 2 in the comments instead of A and B)?

Reply:
you're totally right – will make that change

Reply:
should also try to move everything out of Messages.scala that doesn't need to be there (i.e. isn't part of the sealed Message ADT). That file is getting yuge.